bundles/org.openhab.binding.matter/matter-server/src/bridge/devices/GenericDeviceType.ts

```diff
@@ -0,0 +1,153 @@
+import { FixedLabelServer, LevelControlServer, OnOffServer } from "@matter/main/behaviors";
+import { LevelControl, OnOff } from "@matter/main/clusters";
+import { Logger } from "@matter/main";
+import { TypeFromPartialBitSchema } from "@matter/main/types";
+
+const logger = Logger.get("GenericDevice");
+const OnOffType = OnOffServer.with(OnOff.Feature.Lighting);
+export abstract class GenericDeviceType {
+
+    protected updateLocks = new Set<string>();
+    endpoint: Endpoint;
+    
+
+    constructor(protected bridgeController: BridgeController, protected attributeMap: Record<string, any>, protected endpointId: string,  protected nodeLabel: string, protected productName: string, protected productLabel: string, protected serialNumber: string) {
+        this.nodeLabel = this.#truncateString(nodeLabel);
+        this.productLabel = this.#truncateString(productLabel);
+        this.productName = this.#truncateString(productName);
+        this.serialNumber = this.#truncateString(serialNumber);
+        this.endpoint = this.createEndpoint(this.#generateAttributes(this.defaultClusterValues(), attributeMap));
+        logger.debug(`New Device: label: ${this.nodeLabel} name: ${this.productName} product label: ${this.productLabel} serial: ${this.serialNumber}`);
+    }
+
+    abstract defaultClusterValues(): Record<string, any>;
+    abstract createEndpoint(clusterValues: Record<string, any>): Endpoint;
+
+    public async updateState(clusterName: string, attributeName: string, attributeValue: any) {
+        const args = {} as { [key: string]: any }
+        args[clusterName] = {} as { [key: string]: any }
+        args[clusterName][attributeName] = attributeValue
+        await this.endpoint.set(args);
+    }
+
+    protected sendBridgeEvent(clusterName: string, attributeName: string, attributeValue: any) {
+        const be: BridgeEvent = {
+            type: BridgeEventType.AttributeChanged,
+            data: {
+                endpointId: this.endpoint.id,
+                clusterName: clusterName,
+                attributeName: attributeName,
+                data: attributeValue,
+            }
+        }
+        this.sendEvent(EventType.BridgeEvent, be)
+    }
+
+    protected sendEvent(eventName: string, data: any) {
+        logger.debug(`Sending event: ${eventName} with data: ${data}`);
+        this.bridgeController.ws.sendEvent(eventName, data)
+    }
+
+    protected endPointDefaults() {
+        return {
+            id: this.endpointId,
+            bridgedDeviceBasicInformation: {
+                nodeLabel: this.nodeLabel,
+                productName: this.productName,
+                productLabel: this.productLabel,
+                serialNumber: this.serialNumber,
+                reachable: true,
+            }
+        }
+    }
+    
+    //note that these overrides assume openHAB will be sending the state back when changed as we will not set it here prematurely  
```

Comment:
Couldn't those overrides be part of the subclasses? Having them here feels wrong.

E.g. ColorDeviceType has createColorControlServer ...